Created a Local development topics landing page and made several othe… (#169)
madewithkoji/docs: src/asciidoc/docs/develop/publishLocal.adoc

= Publishing a project you developed locally
:page-slug: publish-locally-developed
:page-description: How to push local changes to the origin repository and pull them into the Koji editor, so you can publish the project.

In the <<work-locally#, previous section>> of this tutorial, you learned how to use Git to clone a Koji project to your development machine and how to install and launch a Node.js server to run your project locally.

In this section, you will
// tag::description[]

> learn how to push the changes that you make locally to your project's origin repository, and how to pull the latest version of your project into the Koji editor, where you can publish it to the world.

Reviewer: Since you no longer need to publish from the editor, we can simplify this intro. E.g.
```suggestion
learn how to push the changes that you make locally to your project's origin repository, so you can publish it to the world.
```

// end::description[]

NOTE: It is recommended that you test the publishing process before you get deep into development locally.
You can make a minor change, push it back to the Koji servers, and check that the change is visible in the published app on Koji.

== Learning objectives

By the end of the section, you should feel comfortable:

* Making and testing changes locally.
* Publishing your changes to the Koji server.
* Pushing your changes to the origin repository, and then pulling them into the Koji editor.

== Prerequisites

* Follow the first two sections to set up the https://withkoji.com/~Svarog1389/rxkd[Falling Objects] game on your development machine.

== Making a change in the source code

Let's make a small change in the https://withkoji.com/~Svarog1389/rxkd[Falling Objects] game, just for testing purposes.

. Open `frontend\src\Components\View\LeaderboardMini.js`.
Change  `<Title>{"Top Players"}</Title>` to `<Title>{"Leaderboard"}</Title>` and save the file.

. In your browser, go to `\http://0.0.0.0:8080`, and make sure the leaderboard shows the new title.

> == Publishing your changes to the Koji server

Reviewer: Since you no longer need to publish from the editor, the flow of the sub-sections in “Publishing your changes to the Koji server” doesn’t quite work. Specially, I would make Testing an h2 and add an h2 that unites the two Koji code editor topics (pulling into editor / updating Node modules). I'll make some inline comments too.

. Open a terminal window and go to the root directory of your project.
. Run the following command.
+
[source,bash]
git push origin master
. At the username prompt, enter your username on Koji and press *Enter*.
. At the password prompt, enter the access key that you <<use-git#_obtaining_an_access_key, generated in a previous tutorial>> and press *Enter*.
+
Your terminal will look something like this:
+
[source,bash]
----
$ git push origin master <1>
Counting objects: 21, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (21/21), done.
Writing objects: 100% (21/21), 84.53 KiB | 0 bytes/s, done.
Total 21 (delta 15), reused 0 (delta 0)
Username for 'https://projects.koji-cdn.com': KojiCoder
Password for 'https://KojiCoder@projects.koji-cdn.com': <2>
To https://projects.koji-cdn.com/a70f8329-e89e-48b0-8d85-7658c1542b9f.git
  a88036c..ea6bda1  master -> master
----
<1> Use Git to upload your new version to the Koji repository
<2> Paste your 32-character access key.

[NOTE]
Pushing your changes to origin automatically publishes a new version of your app.

> === Testing your published app

Reviewer: ```suggestion
== Testing your published app
```

Now that you've published your app, you will want to make sure that it works the same on the Koji server as it does locally.

> 1. Go to the https://withkoji.com/[Koji home page] and sign in to your account.

Reviewer: I think we can remove the login step and assume they know how to log in. We don't have it on the other pages (e.g. cloning and getting project info).

> 2. Click your profile button and go to *Apps > Developer Portal*.

Reviewer: ```suggestion
. On the https://withkoji.com/[Koji home page], click your profile button and go to *Your Apps > Developer Portal*.
```
One way to combine the previous step into this one.

3. Select your app from *All Projects*.

> 4. Click *View Published Koji*.

Reviewer: From the project details, they have a few options that we might want to call out (rather than just View Published Koji). Specifically, it would be nice to add something here about how you can run the debugger for something you are developing locally. E.g., overriding the environment settings. You can just mention it and then link to the debugging page. Also, might be nice to mention that you can verify the status of the build/deploy here.

The published version of your app opens.
5. Proceed to test all the features of your app.

> === Pulling changes into the Koji editor

Reviewer: ```suggestion
== Updating the Koji code editor to match your local environment

If you want to use the Koji code editor after you make changes locally, you must pull your changes into it and update the Node modules to match your local environment.

=== Pulling changes into the Koji editor
```
One way to combine the two sub-sections about how to synchronize the code editor environment with your local environment.

 * branch            master     -> FETCH_HEAD
   0cafbcd..9633a51  master     -> origin/master
Updating 0cafbcd..9633a51
Fast-forward
 frontend/src/Components/Remix/index.js          | 569 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++---------------------------------------------------------------------------------------------------------------------
 frontend/src/Components/View/Leaderboard.js     | 456 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++--------------------------------------------------------------------------------------------
 frontend/src/Components/View/LeaderboardMini.js | 373 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++---------------------------------------------------------------------------
 3 files changed, 673 insertions(+), 725 deletions(-)
root@ip-172-31-12-226:/usr/src/app# 
root@ip-172-31-12-226:/usr/src/app# 
----
+
The code in the Koji editor should now be identical to the code in your local repository.
. In the top right of the Preview pane, click *Refresh* to make sure you are seeing the latest version.
+

> You should now see the custom title, which corresponds to the local change you made in `.koji/customization/strings.json`.

Reviewer: ```suggestion
You should now see the custom title, which corresponds to the local change you made in `frontend\src\Components\View\LeaderboardMini.js`.
```

:page-slug: publish-locally-developed
:page-description: How to push local changes to the origin repository and pull them into the Koji editor, so you can publish the project.

In the previous section of this tutorial, you learned how to use Git to clone a Koji project to your development machine and how to install and launch a Node.js server to run your project locally.

> If you have followed the first two parts of this tutorial, you will now have a scaffold project running on your local computer, so you can use your favorite text editor to modify it and add new features.

Reviewer: This intro feels a bit long and redundant now. Maybe combine some of the concepts into some shorter sentences and/or leave out the restated parts. Let me know if you want to brainstorm.

You will also be able to test that these new features work on your development machine.

In this section, you will
// tag::description[]
learn how to push the changes that you have made locally to your project's origin repository, and how to pull the latest version of your project into the Koji editor, where you can publish it to the world.
// end::description[]

NOTE: It is recommended that you test the publishing process before you get deep into development locally.
You can make a minor change, push it back to the Koji servers, and check that the change is visible in the published template on Koji.

== Learning objectives

> * Learn how to make and test changes locally.

Reviewer: Same thing for the heading and bullets on this page.

* Learn how to push your changes to the origin repository, and then pull them into the Koji editor.
* Learn how to publish your changes to the Koji server.

> == Prerequisites

Reviewer: These prerequisites feel a bit redundant with the intro. Can we assume that the tutorial is completed in order or do we need to call that out as a separate section?

* Install Node.js and launch your project locally.
See the second part of the tutorial.

== Making a change in the source code

Let's make a small change in the https://withkoji.com/~Svarog1389/rxkd[Falling Objects] game, just for testing purposes.

. Open `frontend\src\Components\View\LeaderboardMini.js`.

> Change `<Title>{`Top Players`}</Title>` to `<Title>{`Leaderboard`}</Title>` and save the file.

Reviewer: The formatting is a bit broken. I think the backticks in the code sample are breaking the parsing. You might need to change the character or escape them.

. In your browser, go to `\http://0.0.0.0:8080`, and make sure the leaderboard shows the new title.

> == Moving changes from local to live

Reviewer: Can you double-check with Marko or Sean to see if you still need to publish from the editor or if pushes to origin trigger a full deployment? I can't remember if we changed that or if I just hoped that we did. You might still need to pull changes into the editor if you want to develop there. If so, update these steps and add optional steps for pulling changes into the editor. If not, just resolve this ticket as wishful thinking.

> 

Reviewer: This entire section needs a complete rewrite. There aren't any customization files anymore. "Visual mode" isn't a thing (we have previews now). VCCs are deprecated. Instead, we should be talking about how to use the debugger in conjunction with local development to test things like remix mode or in-app purchases. For some tips and tricks, you can talk to Marko. Most of this section probably just needs to be deleted.
